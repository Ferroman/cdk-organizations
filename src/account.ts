import { CustomResource } from "aws-cdk-lib";
import { AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId } from "aws-cdk-lib/custom-resources";
import { Construct, IConstruct } from "constructs";
import { AccountProvider } from "./account-provider";
import { IParent } from "./parent";
import { IPolicyAttachmentTarget } from "./policy-attachment";
/**
 * @see https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/control-access-billing.html#ControllingAccessWebsite-Activate
 */
export enum IamUserAccessToBilling {
  /**
   * If set to ALLOW, the new account enables IAM users to access account billing information if they have the required permissions.
   */
  ALLOW = "ALLOW",
  /**
   * If set to DENY, only the root user of the new account can access account billing information.
   */
  DENY = "DENY",
}

export interface AccountProps {
  /**
   * The email address of the owner to assign to the new member account. This email address must not already be associated with another AWS account. You must use a valid email address to complete account creation. You can't access the root user of the account or remove an account that was created with an invalid email address.
   */
  readonly email: string;
  /**
   * The friendly name of the member account.
   */
  readonly accountName: string;
  /**
   * The name of an IAM role that AWS Organizations automatically preconfigures in the new member account. This role trusts the management account, allowing users in the management account to assume the role, as permitted by the management account administrator. The role has administrator permissions in the new member account.
   *
   * If you don't specify this parameter, the role name defaults to OrganizationAccountAccessRole.
   */
  readonly roleName?: string;
  /**
   * If set to ALLOW , the new account enables IAM users to access account billing information if they have the required permissions. If set to DENY , only the root user of the new account can access account billing information.
   *
   * @default ALLOW
   */
  readonly iamUserAccessToBilling?: IamUserAccessToBilling;

  /**
   * The parent root or OU that you want to create the new Account in.
   */
  readonly parent?: IParent;
}

export interface IAccount extends IConstruct {
  /**
   * If the account was created successfully, the unique identifier (ID) of the new account. Exactly 12 digits.
   */
  readonly accountId: string;
  /**
   * The friendly name of the account.
   */
  readonly accountName: string;
  /**
   * The email address of the owner to assign to the new member account. This email address must not already be associated with another AWS account. You must use a valid email address to complete account creation. You can't access the root user of the account or remove an account that was created with an invalid email address.
   */
  readonly email: string;
}

/**
 * Creates an AWS account that is automatically a member of the organization whose credentials made the request. AWS Organizations automatically copies the information from the management account to the new member account
 */
export abstract class AccountBase extends Construct implements IAccount, IPolicyAttachmentTarget {
  public abstract readonly accountId: string;
  public abstract readonly accountName: string;
  public abstract readonly email: string;

  identifier(): string {
    return this.accountId;
  }

  currentParentId(): string {
    const parent = new AwsCustomResource(this, "ListParentsCustomResource", {
      onCreate: {
        service: "Organizations",
        action: "listParents", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listParents-property
        region: "us-east-1",
        physicalResourceId: PhysicalResourceId.fromResponse("Parents.0.Id"),
        parameters: {
          ChildId: this.accountId,
        },
      },
      onUpdate: {
        service: "Organizations",
        action: "listParents", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listParents-property
        region: "us-east-1",
        physicalResourceId: PhysicalResourceId.fromResponse("Parents.0.Id"),
        parameters: {
          ChildId: this.accountId,
        },
      },
      onDelete: {
        service: "Organizations",
        action: "listParents", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#listParents-property
        region: "us-east-1",
        physicalResourceId: PhysicalResourceId.fromResponse("Parents.0.Id"),
        parameters: {
          ChildId: this.accountId,
        },
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });

    return parent.getResponseField("Parents.0.Id");
  }

  move(destinationParentId: string, sourceParentId: string): void {
    if (destinationParentId == sourceParentId) {
      return;
    }

    new AwsCustomResource(this, "MoveAccountCustomResource", {
      onUpdate: {
        service: "Organizations",
        action: "moveAccount", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#moveAccount-property
        region: "us-east-1",
        physicalResourceId: PhysicalResourceId.of(this.accountId),
        parameters: {
          AccountId: this.accountId,
          DestinationParentId: destinationParentId,
          SourceParentId: sourceParentId,
        },
      },
      installLatestAwsSdk: false,
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: AwsCustomResourcePolicy.ANY_RESOURCE,
      }),
    });
  }
}

export interface AccountAttributes {
  readonly accountId: string;
  readonly parent: IParent;
}

export class Account extends AccountBase {
  /**
   * Import an existing account from account id.
   */
  public static fromAccountId(scope: Construct, id: string, attrs: AccountAttributes): IAccount {
    class Import extends AccountBase {
      public readonly accountId = attrs.accountId;
      public readonly accountName: string;
      public readonly email: string;

      public constructor() {
        super(scope, id);

        const account = new AwsCustomResource(this, "MoveAccountCustomResource", {
          onCreate: {
            service: "Organizations",
            action: "describeAccount", // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Organizations.html#describeAccount-property
            region: "us-east-1",
            physicalResourceId: PhysicalResourceId.of(this.accountId),
            parameters: {
              AccountId: this.accountId,
            },
          },
          installLatestAwsSdk: false,
          policy: AwsCustomResourcePolicy.fromSdkCalls({
            resources: AwsCustomResourcePolicy.ANY_RESOURCE,
          }),
        });

        this.accountName = account.getResponseField("Account.Name");
        this.email = account.getResponseField("Account.Email");

        if (attrs.parent) {
          const sourceParentId = this.currentParentId();
          const targetParentId = attrs.parent.identifier();
          this.move(targetParentId, sourceParentId);
        }
      }
    }

    return new Import();
  }

  public readonly accountId: string;
  public readonly accountName: string;
  public readonly email: string;

  public constructor(scope: Construct, id: string, props: AccountProps) {
    super(scope, id);

    const { email, accountName, roleName, iamUserAccessToBilling, parent } = props;

    const accountProvider = AccountProvider.getOrCreate(this);
    const account = new CustomResource(this, `AccountProvider-${accountName}`, {
      serviceToken: accountProvider.provider.serviceToken,
      resourceType: "Custom::Organization_Account",
      properties: {
        Email: email,
        AccountName: accountName,
        RoleName: roleName || "OrganizationAccountAccessRole",
        IamUserAccessToBilling: iamUserAccessToBilling || IamUserAccessToBilling.ALLOW,
      },
    });
    this.accountId = account.getAtt("AccountId").toString();
    this.accountName = account.getAtt("AccountName").toString();
    this.email = email;

    if (parent) {
      const sourceParentId = this.currentParentId();
      const targetParentId = parent.identifier();
      this.move(targetParentId, sourceParentId);
    }
  }
}
