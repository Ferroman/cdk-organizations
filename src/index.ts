export * from "./account";
export * from "./enable-aws-service-access";
export * from "./enable-policy-type";
export * from "./delegated-administrator";
export * from "./organization";
export * from "./organizational-unit";
export * from "./parent";
export * from "./policy";
export * from "./policy-attachment";
export * from "./resource";
export * from "./tag-resource";
export * from "./validators";
