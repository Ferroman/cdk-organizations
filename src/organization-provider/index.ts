export * from "./on-event-handler-function";
export * from "./organization-provider";
