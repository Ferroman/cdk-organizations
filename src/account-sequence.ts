import { IAspect, Stack } from "aws-cdk-lib";
import { IConstruct } from "constructs";
import { Account } from "./account";

export class AccountSequence implements IAspect {
  private sequences: { [stackName: string]: Account[] } = {};

  visit(current: IConstruct): void {
    if (!(current instanceof Account)) {
      return;
    }

    const stackName = Stack.of(current).stackName;
    const sequence = this.sequences[stackName];

    if (sequence) {
      const previous = sequence[sequence.length - 1];
      current.node.addDependency(previous);
      sequence.push(current);
    }

    this.sequences[stackName] = [current];
  }
}
