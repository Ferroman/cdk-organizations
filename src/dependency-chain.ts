import { IAspect, Stack } from "aws-cdk-lib";
import { IConstruct } from "constructs";
import { Account } from "./account";
import { DelegatedAdministrator } from "./delegated-administrator";
import { EnableAwsServiceAccess } from "./enable-aws-service-access";
import { EnablePolicyType } from "./enable-policy-type";
import { OrganizationalUnit } from "./organizational-unit";
import { PolicyAttachment } from "./policy-attachment";

/**
 * Aspect to create dependency chain of organization resource that needs to be deployed sequentially
 * @experimental
 */
export class DependencyChain implements IAspect {
  private previous: { [stackName: string]: IConstruct } = {};

  visit(current: IConstruct): void {
    if (!this.needsChaining(current)) {
      return;
    }

    const stackName = Stack.of(current).stackName;

    if (this.previous[stackName]) {
      current.node.addDependency(this.previous[stackName]);
    }

    this.previous[stackName] = current;
  }

  private needsChaining(current: IConstruct): boolean {
    switch (true) {
      case current instanceof EnablePolicyType:
        return true;
      case current instanceof EnableAwsServiceAccess:
        return true;
      case current instanceof DelegatedAdministrator:
        return true;
      case current instanceof Account:
        return true;
      case current instanceof OrganizationalUnit:
        return true;
      case current instanceof PolicyAttachment:
        return true;
      default:
        return false;
    }
  }
}
