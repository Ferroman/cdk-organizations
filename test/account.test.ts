import { Template } from "@aws-cdk/assertions";
import { App, Aspects, Stack } from "@aws-cdk/core";
import { AwsSolutionsChecks } from "cdk-nag";
import { Account, IamUserAccessToBilling } from "../src";

describe("Organization", () => {
  it("Should match snapshot", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "Stack");

    // When
    new Account(stack, "Account", {
      email: "info@pepperize.com",
      accountName: "test",
      tags: {},
    });

    // Then
    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });

  it("Should comply to best practices", () => {
    // Given
    const app = new App();
    const stack = new Stack(app, "Stack");

    // When
    new Account(stack, "Account", {
      email: "info@pepperize.com",
      accountName: "test",
      roleName: "PreconfiguredRoleForNewMemberAccounts",
      iamUserAccessToBilling: IamUserAccessToBilling.DENY,
      tags: {},
    });

    // Then
    Aspects.of(app).add(new AwsSolutionsChecks({ verbose: true }));
  });
});
