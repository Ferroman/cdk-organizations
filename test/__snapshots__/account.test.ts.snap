// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630eArtifactHashD3ACF0DC": Object {
      "Description": "Artifact hash for asset \\"14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630e\\"",
      "Type": "String",
    },
    "AssetParameters14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630eS3BucketFB7E8D66": Object {
      "Description": "S3 bucket for asset \\"14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630e\\"",
      "Type": "String",
    },
    "AssetParameters14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630eS3VersionKey0646B005": Object {
      "Description": "S3 key for asset version \\"14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630e\\"",
      "Type": "String",
    },
    "AssetParameters5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32aArtifactHash28584FC1": Object {
      "Description": "Artifact hash for asset \\"5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32a\\"",
      "Type": "String",
    },
    "AssetParameters5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32aS3Bucket7A285FDA": Object {
      "Description": "S3 bucket for asset \\"5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32a\\"",
      "Type": "String",
    },
    "AssetParameters5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32aS3VersionKeyE54917DE": Object {
      "Description": "S3 key for asset version \\"5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32a\\"",
      "Type": "String",
    },
    "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dArtifactHash6BDDF757": Object {
      "Description": "Artifact hash for asset \\"70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8d\\"",
      "Type": "String",
    },
    "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3BucketF0537962": Object {
      "Description": "S3 bucket for asset \\"70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8d\\"",
      "Type": "String",
    },
    "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3VersionKey46386A78": Object {
      "Description": "S3 key for asset version \\"70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8d\\"",
      "Type": "String",
    },
    "AssetParameters7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cdArtifactHashF658FD15": Object {
      "Description": "Artifact hash for asset \\"7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cd\\"",
      "Type": "String",
    },
    "AssetParameters7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cdS3BucketA00C8555": Object {
      "Description": "S3 bucket for asset \\"7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cd\\"",
      "Type": "String",
    },
    "AssetParameters7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cdS3VersionKey27C92598": Object {
      "Description": "S3 key for asset version \\"7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cd\\"",
      "Type": "String",
    },
    "AssetParametersbda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532ArtifactHashAEF3E82B": Object {
      "Description": "Artifact hash for asset \\"bda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532\\"",
      "Type": "String",
    },
    "AssetParametersbda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532S3BucketE75A9764": Object {
      "Description": "S3 bucket for asset \\"bda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532\\"",
      "Type": "String",
    },
    "AssetParametersbda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532S3VersionKeyE22D3120": Object {
      "Description": "S3 key for asset version \\"bda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": Object {
      "DependsOn": Array [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3BucketF0537962",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3VersionKey46386A78",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters70893b631249dc61260989e92e90d60ae94fbbec490a1e065680d77383084d8dS3VersionKey46386A78",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AccountAccountProvidertestE06C558F": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccountName": "test",
        "Email": "info@pepperize.com",
        "IamUserAccessToBilling": "ALLOW",
        "RoleName": "OrganizationAccountAccessRole",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AccountpepperizecdkorganizationsAccountProviderNestedStackpepperizecdkorganizationsAccountProviderNestedStackResource05676357",
            "Outputs.StackAccountpepperizecdkorganizationsAccountProviderframeworkonEventFDA17817Arn",
          ],
        },
      },
      "Type": "Custom::Organization_Account",
      "UpdateReplacePolicy": "Delete",
    },
    "AccountListParentsCustomResource5959DA0B": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "AccountListParentsCustomResourceCustomResourcePolicy508C2A1A",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listParents\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Parents.0.Id\\"},\\"parameters\\":{\\"ChildId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountAccountProvidertestE06C558F",
                  "AccountId",
                ],
              },
              "\\"}}",
            ],
          ],
        },
        "Delete": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listParents\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Parents.0.Id\\"},\\"parameters\\":{\\"ChildId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountAccountProvidertestE06C558F",
                  "AccountId",
                ],
              },
              "\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listParents\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Parents.0.Id\\"},\\"parameters\\":{\\"ChildId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountAccountProvidertestE06C558F",
                  "AccountId",
                ],
              },
              "\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "AccountListParentsCustomResourceCustomResourcePolicy508C2A1A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "organizations:ListParents",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AccountListParentsCustomResourceCustomResourcePolicy508C2A1A",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AccountMoveAccountCustomResourceCustomResourcePolicy09414514": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "organizations:MoveAccount",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AccountMoveAccountCustomResourceCustomResourcePolicy09414514",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AccountMoveAccountCustomResourceD4C42319": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "AccountMoveAccountCustomResourceCustomResourcePolicy09414514",
      ],
      "Properties": Object {
        "Create": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"Organizations\\",\\"action\\":\\"moveAccount\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"id\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountAccountProvidertestE06C558F",
                  "AccountId",
                ],
              },
              "\\"},\\"parameters\\":{\\"AccountId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountAccountProvidertestE06C558F",
                  "AccountId",
                ],
              },
              "\\",\\"DestinationParentId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "OrganizationRootRootCustomResourceBB74F060",
                  "Roots.0.Id",
                ],
              },
              "\\",\\"SourceParentId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountListParentsCustomResource5959DA0B",
                  "Parents.0.Id",
                ],
              },
              "\\"}}",
            ],
          ],
        },
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"service\\":\\"Organizations\\",\\"action\\":\\"moveAccount\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"id\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountAccountProvidertestE06C558F",
                  "AccountId",
                ],
              },
              "\\"},\\"parameters\\":{\\"AccountId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountAccountProvidertestE06C558F",
                  "AccountId",
                ],
              },
              "\\",\\"DestinationParentId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "OrganizationRootRootCustomResourceBB74F060",
                  "Roots.0.Id",
                ],
              },
              "\\",\\"SourceParentId\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AccountListParentsCustomResource5959DA0B",
                  "Parents.0.Id",
                ],
              },
              "\\"}}",
            ],
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "AccountpepperizecdkorganizationsAccountProviderNestedStackpepperizecdkorganizationsAccountProviderNestedStackResource05676357": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Parameters": Object {
          "referencetoStackAssetParameters5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32aS3Bucket0C81A20ARef": Object {
            "Ref": "AssetParameters5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32aS3Bucket7A285FDA",
          },
          "referencetoStackAssetParameters5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32aS3VersionKey04813294Ref": Object {
            "Ref": "AssetParameters5c356b83dcca0c06053bcacef59d86c46fea6468e254ec901c9824a548b5d32aS3VersionKeyE54917DE",
          },
          "referencetoStackAssetParameters7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cdS3Bucket7A09F772Ref": Object {
            "Ref": "AssetParameters7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cdS3BucketA00C8555",
          },
          "referencetoStackAssetParameters7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cdS3VersionKeyFCF7B7A5Ref": Object {
            "Ref": "AssetParameters7ef22dac3421fb2d25175b3386053dbc5bdc110e87cbd58600263bf70a6324cdS3VersionKey27C92598",
          },
          "referencetoStackAssetParametersbda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532S3BucketA44BD36BRef": Object {
            "Ref": "AssetParametersbda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532S3BucketE75A9764",
          },
          "referencetoStackAssetParametersbda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532S3VersionKey0E435F65Ref": Object {
            "Ref": "AssetParametersbda532d8740e1ee5114c8af882fed3c0ba27d4588f10b637aac97e07919ee532S3VersionKeyE22D3120",
          },
        },
        "TemplateURL": Object {
          "Fn::Join": Array [
            "",
            Array [
              "https://s3.",
              Object {
                "Ref": "AWS::Region",
              },
              ".",
              Object {
                "Ref": "AWS::URLSuffix",
              },
              "/",
              Object {
                "Ref": "AssetParameters14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630eS3BucketFB7E8D66",
              },
              "/",
              Object {
                "Fn::Select": Array [
                  0,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630eS3VersionKey0646B005",
                      },
                    ],
                  },
                ],
              },
              Object {
                "Fn::Select": Array [
                  1,
                  Object {
                    "Fn::Split": Array [
                      "||",
                      Object {
                        "Ref": "AssetParameters14c7461124acf258f19d782bdfc67b035f7c49c468766a0c57ce2ad7dc58630eS3VersionKey0646B005",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::CloudFormation::Stack",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationOrganizationCustomResource620D04AC": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationOrganizationCustomResourceCustomResourcePolicy72E5C59D",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"Organization\\",\\"action\\":\\"createOrganization\\",\\"region\\":\\"us-east-1\\",\\"parameters\\":{\\"FeatureSet\\":\\"ALL\\"},\\"physicalResourceId\\":{\\"responsePath\\":\\"Organization.Id\\"}}",
        "Delete": "{\\"service\\":\\"Organization\\",\\"action\\":\\"deleteOrganization\\",\\"region\\":\\"us-east-1\\"}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"Organization\\",\\"action\\":\\"describeOrganization\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Organization.Id\\"}}",
      },
      "Type": "Custom::Organization_Organization",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationOrganizationCustomResourceCustomResourcePolicy72E5C59D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "organization:CreateOrganization",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "organization:DescribeOrganization",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "organization:DeleteOrganization",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OrganizationOrganizationCustomResourceCustomResourcePolicy72E5C59D",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "OrganizationRootRootCustomResourceBB74F060": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "OrganizationOrganizationCustomResourceCustomResourcePolicy72E5C59D",
        "OrganizationOrganizationCustomResource620D04AC",
        "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
      ],
      "Properties": Object {
        "Create": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Roots.0.Id\\"}}",
        "Delete": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\"}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{\\"service\\":\\"Organizations\\",\\"action\\":\\"listRoots\\",\\"region\\":\\"us-east-1\\",\\"physicalResourceId\\":{\\"responsePath\\":\\"Roots.0.Id\\"}}",
      },
      "Type": "Custom::Organization_Root",
      "UpdateReplacePolicy": "Delete",
    },
    "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E": Object {
      "DependsOn": Array [
        "OrganizationOrganizationCustomResourceCustomResourcePolicy72E5C59D",
        "OrganizationOrganizationCustomResource620D04AC",
        "OrganizationRootRootCustomResourceBB74F060",
      ],
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "organizations:ListRoots",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "OrganizationRootRootCustomResourceCustomResourcePolicyB45F831E",
        "Roles": Array [
          Object {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
